pkg/k8sops/types.go

```diff
@@ -0,0 +1,128 @@
+
+	// GenerateService will produce resource configured according to the spec's
+	// serviceConfiguration fields
+	GenerateService(svcCfg myspec.ServiceConfiguration) *v1.Service
+
+	// GenerateServicePorts will produce the correct ServicePort or ContainerPort
+	// resources
+	GenerateServicePorts(ports []myspec.Port) []v1.ServicePort
+
+	// GenerateContainerPorts will produce ServicePorts given a
+	// ServiceConfiguration
+	GenerateContainerPorts(ports []myspec.Port) []v1.ContainerPort
+
+	// MonitorM3DBEvents watches for new or removed clusters
+	MonitorM3DBEvents(stopchan chan struct{}) (<-chan *myspec.M3DBCluster, <-chan error)
+
+	// GetService simply gets a service by name
+	GetService(cluster *myspec.M3DBCluster, name string) (*v1.Service, error)
+
+	// DeleteService simply deletes a service by name
+	DeleteService(cluster *myspec.M3DBCluster, name string) error
+
+	// EnsureService will create a service by name if it doesn't exist
+	EnsureService(cluster *myspec.M3DBCluster, svcCfg myspec.ServiceConfiguration) error
+
+	// MultiLabelSelector provides a ListOptions with a LabelSelector
+	// given a map of strings
+	MultiLabelSelector(kvs map[string]string) metav1.ListOptions
+
+	// LabelSelector provides a ListOptions with a LabelSelector given a key
+	// and value strings
+	LabelSelector(key, value string) metav1.ListOptions
+
+	// DeleteStatefulSets will delete all stateful sets associated with a cluster
+	DeleteStatefulSets(cluster *myspec.M3DBCluster, listOpts metav1.ListOptions) error
+
+	// GetStatefulSets provides all the StatefulSets contained within a
+	// cluster
+	GetStatefulSets(cluster *myspec.M3DBCluster, listOpts metav1.ListOptions) (*appsv1.StatefulSetList, error)
+
+	// GetPlacementDetails provides the pod to isolation group mapping
+	GetPlacementDetails(cluster *myspec.M3DBCluster) (map[string]string, error)
+
+	// GetPodsByLabel provides a PodList given ListOptions which contain the
+	// correct LabelSelector
+	GetPodsByLabel(cluster *myspec.M3DBCluster, listOpts metav1.ListOptions) (*v1.PodList, error)
+
+	// EnsureStatefulSets will create StatefulSets based the Spec configuration
+	// if they don't exist already
+	EnsureStatefulSets(cluster *myspec.M3DBCluster, svcCfg myspec.ServiceConfiguration) error
+
+	// CreateStatefulSet will create a StatefulSet and ensure all Pod replicas are
+	// ready before returning
+	CreateStatefulSet(cluster *myspec.M3DBCluster, statefulSet *appsv1.StatefulSet) (*appsv1.StatefulSet, error)
+
+	// GetStatefulSet simply returns a StatefulSet given the current cluster
+	GetStatefulSet(cluster *myspec.M3DBCluster, name string) (*appsv1.StatefulSet, error)
+
+	// UpdateStatefulSet simply updates a statefulset
+	UpdateStatefulSet(cluster *myspec.M3DBCluster, statefulSet *appsv1.StatefulSet) (*appsv1.StatefulSet, error)
+
+	// CheckStatefulStatus will poll a given StatefulSet to ensure it reaches a
+	// ready state within 60 seconds with polling updates at 500 ms
```

Comment:
probably best to avoid specific numbers here and say "within a configurable period" or something